### ISSUE TITLE
Might want to include some docs on how to change the HashMap implementation

### ISSUE THREAD

**Issue by kimono-koans** [rayon-rs/rayon#953]
Changing the Hasher for HashMap was a little arduous for me, and might be for others.  You might wish to include a small code example somewhere in the [docs](https://docs.rs/rayon/latest/rayon/collections/hash_map/index.html), like so:

```bash
// rayon requires impl AHashMap as (K,V) instead of (K,V,S) so we
// may use with collect() function, see rayon::FromParallelIterator
use ahash::AHasher;
pub type AHashBuildHasher = BuildHasherDefault<AHasher>;
pub type AHashMapSpecial<K, V> = std::collections::HashMap<K, V, AHashBuildHasher>;
use AHashMapSpecial as HashMap;
```

Thanks!  Very much appreciate all your hard work.

**Comment by cuviper** [comment]
I don't understand your explanation, but I also don't really understand where you're coming from, or what was confusing about `HashMap` *specifically* as it may be used with `rayon`. The custom `S` type parameter is not unique to `rayon`.

The signature of [`FromParallelIterator for HashMap`](https://docs.rs/rayon/latest/rayon/iter/trait.FromParallelIterator.html#impl-FromParallelIterator%3C(K%2C%20V)%3E-for-HashMap%3CK%2C%20V%2C%20S%3E) is intentionally similar to [`FromIterator for HashMap`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#impl-FromIterator%3C(K%2C%20V)%3E-1) in the standard library, only with the addition of `Send` constraints.

**Comment by kimono-koans** [comment]
My experience has been that many custom HashMap wrappers for Hasher crates don't work with `rayon`.  And the error message one gets, in hindsight, may make some sense, but the difference between the impl for the stdlib HashMap and the ahash HashMap wrapper wasn't immediately clear.  See the error: 

```
error[E0277]: the trait bound `ahash::AHashMap<std::path::PathBuf, std::vec::Vec<std::path::PathBuf>>: rayon::iter::FromParallelIterator<(std::path::PathBuf, std::vec::Vec<std::path::PathBuf>)>` is not satisfied
    --> src/parse_mounts.rs:266:10
     |
266  |         .collect()
     |          ^^^^^^^ the trait `rayon::iter::FromParallelIterator<(std::path::PathBuf, std::vec::Vec<std::path::PathBuf>)>` is not implemented for `ahash::AHashMap<std::path::PathBuf, std::vec::Vec<std::path::PathBuf>>`
     |
     = help: the following other types implement trait `rayon::iter::FromParallelIterator<T>`:
               <() as rayon::iter::FromParallelIterator<()>>
               <(A, B) as rayon::iter::FromParallelIterator<itertools::Either<L, R>>>
               <(FromA, FromB) as rayon::iter::FromParallelIterator<(A, B)>>
               <std::borrow::Cow<'a, C> as rayon::iter::FromParallelIterator<T>>
               <std::collections::BTreeMap<K, V> as rayon::iter::FromParallelIterator<(K, V)>>
               <std::collections::BTreeSet<V> as rayon::iter::FromParallelIterator<V>>
               <std::collections::BinaryHeap<T> as rayon::iter::FromParallelIterator<T>>
               <std::collections::HashMap<K, V, S> as rayon::iter::FromParallelIterator<(K, V)>>
             and 11 others
note: required by a bound in `rayon::iter::ParallelIterator::collect`
    --> /Users/rswinford/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.3/src/iter/mod.rs:2044:12
     |
2044 |         C: FromParallelIterator<Self::Item>,
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `rayon::iter::ParallelIterator::collect`

```

Changing the Hasher on the default std HashMap is very common.  Using a HashMap wrapper from a Hasher crate is also very, very common.  And use with `rayon` may be the first time a new user confronts this particular issue.  It's perfectly true I'm a big dummy, and I should have figured this out sooner, but with an eye towards helping a new user, and since I could see this happening with *great frequency*, I thought a small blurb in the docs about how to do this relatively common thing might helpful.

On the other hand, I can perfectly understand if you don't think this merits such a blurb.  Not going to break my heart if you close.  This was just a sharp edge for me and it took me a second to figure out.

Thanks.

**Comment by adamreichold** [comment]
So the problem with `AHashMap` is that it is not just a special parameterization of std's `HashMap`, but a distinct type that albeit just wraps std's `HashMap`.

I think from the point of view of Rayon, this is similar to any other container implementation which needs to provide its own implementation of Rayon's traits as these implementations usually depend on internals of the crate. This is usually done by having a `rayon` feature exposed by the container crates themselves, c.f. [`hashbrown`](https://docs.rs/crate/hashbrown/latest/features).

I think the best way forward here would be to make `AHashMap` compatible with `rayon` by adding such a feature even that the implementation would in this case only forward to Rayon's own implementation for the std types.

**Comment by kimono-koans** [comment]
Good idea.  I'll file an issue and link this issue with `ahash`.  Also it may be enough for some user that they are able my/this issue.  Thanks.
